File: tests/test_opt_sample_cap.py
"""opt_sample_cap: the classical Grover-round sample cap (bd 0o8).

A faithful CBQS solve simulates a Grover round of j iterations by drawing 4j^2+1
candidate states (CSearch_{sat,opt_sat,opt}); that O(n*j^2) classical cost makes
large-n (large-j) solves intractable and blocks the large-n baseline freeze.
`opt_sample_cap` bounds a round to `cap` candidates. It must NOT change the
oracle accounting -- the 2j+1 charge is applied in ctg BEFORE the sim
(CLAUDE.md sec 1.2) -- so the per-worker budget T(n) still binds and single-worker
runs stay deterministic. cap == 0 (default) is the exact unbounded sim, so every
existing baseline stays green.

These are the known-correct gates (not no-crash checks):
  * a cap larger than any round's realized sample count is BIT-IDENTICAL to cap=0
    (proves the cap path does not perturb the PRNG stream or the accounting);
  * a binding cap on a stalling instance still terminates at exactly T(n);
  * a capped run is deterministic and feasibility-correct (verify=True).
"""

import pytest

pytest.importorskip("cbqs")

from cbqs.Model import Model
from cbqs.Constants import MAXIMIZE


def _knapsack(n):
    """n-variable knapsack that improves several times before converging."""
    m = Model()
    x = m.add_variables(n)
    m.set_objective(sum((i % 7 + 1) * x[i] for i in range(n)), sense=MAXIMIZE)
    m.add_constraint(sum((i % 5 + 1) * x[i] for i in range(n)) <= n * 2)
    m.close()
    return m


def _stalling(n):
    """Covering (>=) knapsack that dwells in long non-improving stretches, so
    `rounds` grows and a stage-3 Grover round draws a large j -> the 4j^2+1 sim
    blows up. This is exactly the regime opt_sample_cap is meant to bound
    (mirrors test_oracle_budget._stalling)."""
    m = Model()
    x = m.add_variables(n)
    w = [i % 5 + 1 for i in range(n)]
    m.set_objective(sum((i % 7 + 1) * x[i] for i in range(n)), sense=MAXIMIZE)
    m.add_constraint(sum(w[i] * x[i] for i in range(n)) >= sum(w) - 2)
    m.close()
    return m


def _T(n):
    """Default per-worker oracle budget T(n) = (n/4)^2 + 1200 (NORTHSTAR sec 3)."""
    return int((n / 4.0) ** 2 + 1200)


def _solve(mk, n, cap, *, seed=7, M=-1, workers=1, verify=False):
    m = mk(n)
    m.seed = seed
    m.set_param("M", M)
    m.set_param("num_workers", workers)
    m.set_param("opt_sample_cap", cap)
    if verify:
        m.set_param("verify", True)
    return m.solve()


class TestOptSampleCap:
    def test_default_is_zero_unbounded(self):
        """The default is 0 (exact unbounded sim) -- the cap is opt-in, so no
        committed baseline is affected unless a finite cap is set."""
        m = _knapsack(20)
        assert m._get_effective("opt_sample_cap") == 0

    def test_negative_cap_rejected(self):
        """The param guard rejects a negative cap (domain is >= 0)."""
        m = _knapsack(20)
        with pytest.raises((ValueError, Exception)):
            m.set_param("opt_sample_cap", -1)

    def test_large_cap_is_bit_identical_to_unbounded(self):
        """A cap larger than any round's realized sample count must reproduce the
        unbounded run BIT-FOR-BIT: identical oracle_calls, objective, AND solution
        array. This proves the cap code path -- when it does not bind -- neither
        perturbs the per-worker PRNG stream nor the oracle accounting (CLAUDE.md
        sec 1.2 / sec 8). n=40 keeps every round's 4j^2+1 well under 10**9."""
        r0 = _solve(_knapsack, 40, 0)
        rC = _solve(_knapsack, 40, 10**9)
        assert r0.oracle_calls == rC.oracle_calls
        assert r0.objective == rC.objective
        assert list(r0.solution) == list(rC.solution)

    def test_cap_preserves_oracle_budget_on_stalling(self):
        """REGRESSION: capping the classical sim must NOT change the oracle
        budget. A stalling instance (which forces large j) still terminates at
        exactly T(n) with a binding cap, just as the uncapped run does -- the
        2j+1 charge and the cumulative-budget gate in ctg are untouched."""
        n = 60
        T = _T(n)
        r_unc = _solve(_stalling, n, 0)
        r_cap = _solve(_stalling, n, 300)
        assert T - 2 <= r_unc.oracle_calls <= T
        assert T - 2 <= r_cap.oracle_calls <= T, (
            f"capped stalling oracle_calls={r_cap.oracle_calls} not in "
            f"[{T-2}, {T}] -- the cap must not alter the oracle budget"
        )

    def test_capped_run_is_deterministic(self):
        """Fixed seed + single worker + a binding cap => identical oracle_calls,
        objective, AND solution across runs (determinism survives the cap, sec 8)."""
        r1 = _solve(_stalling, 60, 300)
        r2 = _solve(_stalling, 60, 300)
        assert r1.oracle_calls == r2.oracle_calls
        assert r1.objective == r2.objective
        assert list(r1.solution) == list(r2.solution)

    def test_capped_run_is_feasible(self):
        """verify=True asserts every reported-feasible solution satisfies
        eval_constraints==0 (Model.pyx). A binding cap must not poison
        feasibility/sign accounting (CLAUDE.md sec 2.1 / sec 5)."""
        r = _solve(_stalling, 80, 500, verify=True)
        assert r.solution is not None

    def test_cap_preserves_multi_worker_budget(self):
        """With a cap and num_workers>1 the reported per-worker oracle_calls is
        still ~T(n) (best-of-portfolio), not corrupted by the cap."""
        n = 40
        T = _T(n)
        r = _solve(_knapsack, n, 200, workers=4)
        assert T <= r.oracle_calls < 2 * T, (
            f"multi-worker capped oracle_calls={r.oracle_calls} should be ~T(n)={T}"
        )


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/test_opt_sample_cap.c
/*
 * test_opt_sample_cap.c -- bd 0o8: the classical Grover-round sample cap.
 *
 * CSearch_{sat,opt_sat,opt} faithfully simulate a Grover round of j iterations
 * by drawing 4j^2+1 candidate states and returning the first improver. That sim
 * is O(n*j^2) per round, which makes large-n (large-j) solves intractable.
 * ctx->opt_sample_cap > 0 bounds the count to `cap` candidates WITHOUT touching
 * the 2j+1 oracle charge (which lives in ctg, before CSearch_*). These tests pin:
 *   1. cap == 0 reproduces the exact legacy count 4j^2+1 (faithfulness, §8).
 *   2. cap > 0 bounds the realized candidate count to min(4j^2+1, cap).
 *   3. a large j where the legacy `int` expression 4*j*j+1 would overflow
 *      (j > ~23170) is handled correctly under int64 + cap (latent-UB fix).
 *
 * ctx->opt_candidates increments exactly once per generated candidate
 * (solver.c, pure instrumentation), so it is the observable candidate count.
 * We drive an already-optimal incumbent so CSearch_opt never early-returns and
 * therefore runs the full (capped) sample budget.
 */
#include <stdarg.h>
#include <stddef.h>
#include <setjmp.h>
#include <cmocka.h>

#include "solver.h"
#include "model.h"
#include "constraint.h"
#include "Expression.h"
#include "Branching.h"
#include "solver_ctx.h"
#include "definitions.h"
#include "prng.h"

/* x0 + x1 + x2 <= 0, objective -(x0+x1+x2) MINIMIZE. The constraint forces the
 * UNIQUE feasible point (0,0,0) (every variable is pinned to 0 by look-ahead),
 * so the incumbent (0,0,0)/tot_profit=0 is unimprovable: CSearch_opt never
 * early-returns and runs the full (capped) sample budget -- making
 * ctx->opt_candidates a clean, deterministic readout of the realized count. */
static model_t *build_small_model(void) {
    model_t *mod = init_model();

    expression_t *con_expr = init_expression();
    add_variable(con_expr, 0);
    add_variable(con_expr, 1);
    add_variable(con_expr, 2);
    add_sense_to_expression(con_expr, LOWER);
    add_rhs_to_expression(con_expr, 0);
    add_expression_to_constraints(mod->con, con_expr);

    expression_t *obj_expr = init_expression();
    add_variable(obj_expr, 0);
    multiply_constant(obj_expr, -1);
    add_variable(obj_expr, 1);
    multiply_constant(obj_expr, -1);
    add_variable(obj_expr, 2);
    multiply_constant(obj_expr, -1);
    add_sense_to_expression(obj_expr, LOWER);
    add_rhs_to_expression(obj_expr, 0);
    add_expression_to_constraints(mod->obj, obj_expr);

    preprocessing(3, mod->con);
    preprocessing(3, mod->obj);

    int arr[3] = {0, 0, 0};
    mod->initial_state = init_state(0, arr, 3);
    mod->initial_state->tot_profit = INT64_MAX;
    mod->global_opt = init_state(0, arr, 3);
    mod->global_opt->tot_profit = INT64_MAX;

    mod->n = 3;
    mod->M = 10;
    mod->depth_look_ahead = 0;
    mod->stopping_time = 100;
    mod->solver = OPTIMIZE;
    mod->break_item = 0;

    free_expression(con_expr);
    free_expression(obj_expr);
    return mod;
}

/* Run CSearch_opt once on the unimprovable incumbent (0,0,0)=>tot_profit 0 (the
 * unique feasible point under x0+x1+x2<=0), so the loop never early-returns and
 * opt_candidates equals the realized (capped) sample budget. Returns that count. */
static uint64_t run_opt_count(int j, int64_t cap) {
    model_t *mod = build_small_model();
    solver_ctx_t *ctx = solver_ctx_create();
    assert_non_null(ctx);
    ctx->seed = 42;
    solver_ctx_init_prng(ctx);
    ctx->opt_sample_cap = cap;

    int arr[3] = {0, 0, 0};
    state_t *sol = init_state(0, arr, 3);
    sol->tot_profit = 0;    /* unique feasible point: no candidate can beat it */
    sol->feasible = 1;
    int samples = 0;

    int res CBQS_UNUSED = CSearch_opt(ctx, sol, j, mod->con, mod->obj, 0, 1, NULL, &samples);
    assert_int_equal(res, 0);  /* no improvement on an optimal incumbent */

    uint64_t count = ctx->opt_candidates;
    free_state(sol, 1);
    solver_ctx_free(ctx);
    free_model(mod);
    return count;
}

/* cap == 0 reproduces the exact legacy budget 4j^2+1 (bit-for-bit faithful). */
static void test_cap_zero_is_legacy_count(void **state) {
    (void)state;
    assert_int_equal((int) run_opt_count(1, 0), 4 * 1 * 1 + 1);    /* 5   */
    assert_int_equal((int) run_opt_count(5, 0), 4 * 5 * 5 + 1);    /* 101 */
    assert_int_equal((int) run_opt_count(10, 0), 4 * 10 * 10 + 1); /* 401 */
}

/* cap > 0 bounds the count to min(4j^2+1, cap). */
static void test_cap_bounds_candidate_count(void **state) {
    (void)state;
    /* cap binds: 4*10^2+1 = 401 > 50 -> exactly 50 candidates. */
    assert_int_equal((int) run_opt_count(10, 50), 50);
    /* cap does not bind: 4*5^2+1 = 101 < 1000 -> full 101. */
    assert_int_equal((int) run_opt_count(5, 1000), 101);
    /* cap == 4j^2+1 exactly -> full count (boundary). */
    assert_int_equal((int) run_opt_count(5, 101), 101);
}

/* Large j where the legacy `int` 4*j*j+1 overflows (j>~23170): under int64 +
 * cap the loop bound is correct, so a small cap yields exactly `cap` candidates
 * and the call returns quickly instead of UB (negative/wrapped bound -> 0 iters,
 * or a multi-billion-iteration hang). */
static void test_cap_large_j_no_overflow(void **state) {
    (void)state;
    /* 4*30000^2+1 = 3.6e9 overflows int32; cap=100 must bound it to 100. */
    assert_int_equal((int) run_opt_count(30000, 100), 100);
    /* Even at INT_MAX-scale j the cap binds (no overflow, no hang). */
    assert_int_equal((int) run_opt_count(2000000, 7), 7);
}

int main(void) {
    const struct CMUnitTest tests[] = {
        cmocka_unit_test(test_cap_zero_is_legacy_count),
        cmocka_unit_test(test_cap_bounds_candidate_count),
        cmocka_unit_test(test_cap_large_j_no_overflow),
    };
    return cmocka_run_group_tests(tests, NULL, NULL);
}
